### PR TITLE
Add DC-ASGD optimizer

### PR DESCRIPTION
**Pull request by kernel8liang**
1. according to paper "Asynchronous Stochastic Gradient Descent with Delay Compensation for
   Distributed Deep Learning", DC-ASGD nearly approaches the performance of sequential SGD,  
   so implement it. 
2. add "--optimizer" argument in train_minst.py.
3. change metrics to "accuracy"

test train_minst with dockers in same machine.

11 dockers 11 workers bsp sgd with lr 0.1

```
    2016-10-25 05:58:30,787 Node[9] Epoch[9] Validation-accuracy=0.985491
    2016-10-25 05:58:30,787 Node[9] Epoch[9] Validation-accuracy=0.985491
    2016-10-25 05:58:30,789 Node[8] Epoch[9] Validation-accuracy=0.991071
    2016-10-25 05:58:30,789 Node[8] Epoch[9] Validation-accuracy=0.991071
    2016-10-25 05:58:30,796 Node[7] Epoch[9] Validation-accuracy=0.982143
    2016-10-25 05:58:30,796 Node[7] Epoch[9] Validation-accuracy=0.982143
    2016-10-25 05:58:30,799 Node[4] Epoch[9] Validation-accuracy=0.950893
    2016-10-25 05:58:30,800 Node[4] Epoch[9] Validation-accuracy=0.950893
    2016-10-25 05:58:30,808 Node[5] Epoch[9] Validation-accuracy=0.983259
    2016-10-25 05:58:30,808 Node[5] Epoch[9] Validation-accuracy=0.983259
    2016-10-25 05:58:30,810 Node[2] Epoch[9] Validation-accuracy=0.941964
    2016-10-25 05:58:30,810 Node[2] Epoch[9] Validation-accuracy=0.941964
    2016-10-25 05:58:30,811 Node[3] Epoch[9] Validation-accuracy=0.966518
    2016-10-25 05:58:30,811 Node[3] Epoch[9] Validation-accuracy=0.966518
    2016-10-25 05:58:30,817 Node[1] Epoch[9] Validation-accuracy=0.947545
    2016-10-25 05:58:30,817 Node[1] Epoch[9] Validation-accuracy=0.947545
    2016-10-25 05:58:30,816 Node[6] Epoch[9] Validation-accuracy=0.965402
    2016-10-25 05:58:30,816 Node[6] Epoch[9] Validation-accuracy=0.965402
    2016-10-25 05:58:30,827 Node[10] Epoch[9] Validation-accuracy=0.973214
    2016-10-25 05:58:30,828 Node[10] Epoch[9] Validation-accuracy=0.973214
    2016-10-25 05:58:30,830 Node[0] Epoch[9] Validation-accuracy=0.973214
    2016-10-25 05:58:30,830 Node[0] Epoch[9] Validation-accuracy=0.973214
```

11 dockers 11 workers asp dcasgd with lr 0.006

```
    2016-10-25 06:07:55,764 Node[10] Epoch[9] Validation-accuracy=0.966518
    2016-10-25 06:07:55,764 Node[10] Epoch[9] Validation-accuracy=0.966518
    2016-10-25 06:07:58,675 Node[1] Epoch[9] Validation-accuracy=0.940848
    2016-10-25 06:07:58,676 Node[1] Epoch[9] Validation-accuracy=0.940848
    2016-10-25 06:08:00,159 Node[8] Epoch[9] Validation-accuracy=0.985491
    2016-10-25 06:08:00,159 Node[8] Epoch[9] Validation-accuracy=0.985491
    2016-10-25 06:08:04,189 Node[7] Epoch[9] Validation-accuracy=0.982143
    2016-10-25 06:08:04,189 Node[7] Epoch[9] Validation-accuracy=0.982143
    2016-10-25 06:08:04,191 Node[6] Epoch[9] Validation-accuracy=0.972098
    2016-10-25 06:08:04,191 Node[6] Epoch[9] Validation-accuracy=0.972098
    2016-10-25 06:08:04,364 Node[9] Epoch[9] Validation-accuracy=0.983259
    2016-10-25 06:08:04,364 Node[9] Epoch[9] Validation-accuracy=0.983259
    2016-10-25 06:08:04,366 Node[2] Epoch[9] Validation-accuracy=0.946429
    2016-10-25 06:08:04,367 Node[2] Epoch[9] Validation-accuracy=0.946429
    2016-10-25 06:08:04,643 Node[3] Epoch[9] Validation-accuracy=0.966518
    2016-10-25 06:08:04,643 Node[3] Epoch[9] Validation-accuracy=0.966518
    2016-10-25 06:08:05,182 Node[0] Epoch[9] Validation-accuracy=0.968750
    2016-10-25 06:08:05,183 Node[0] Epoch[9] Validation-accuracy=0.968750
    2016-10-25 06:08:05,548 Node[4] Epoch[9] Validation-accuracy=0.947545
    2016-10-25 06:08:05,548 Node[4] Epoch[9] Validation-accuracy=0.947545
    2016-10-25 06:08:05,824 Node[5] Epoch[9] Validation-accuracy=0.979911
    2016-10-25 06:08:05,824 Node[5] Epoch[9] Validation-accuracy=0.979911
```

### CODE REVIEW

File: example/image-classification/train_model.py [hunk 1]
```diff
@@ -37,10 +37,6 @@ def fit(args, network, data_loader, batch_end_callback=None):
         model_args = {'arg_params' : tmp.arg_params,
                       'aux_params' : tmp.aux_params,
                       'begin_epoch' : args.load_epoch}
-        # TODO: check epoch_size for 'dist_sync'
```
Comment (piiswrong) [inline example/image-classification/train_model.py]: please revert changes in this file

File: python/mxnet/optimizer.py [hunk 2]
```diff
@@ -308,6 +309,99 @@ def update(self, index, weight, grad, state):
             assert self.momentum == 0.0
             weight[:] += -lr * (grad + wd * weight)
 
+@register
+class DCASGD(Optimizer):
+    """ DCASGD optimizer with momentum and weight regularization.
```
Comment (piiswrong) [inline python/mxnet/optimizer.py]: Could you add reference to the paper here?

File: python/mxnet/optimizer.py [hunk 3]
```diff
@@ -7,6 +7,7 @@
 from .base import OptimizerHandle, OptimizerCreator
 from .ndarray import NDArray, zeros, clip, sqrt, square
 from .random import normal
+import logging
```
Comment (piiswrong) [inline python/mxnet/optimizer.py]: not used?

**Comment by piiswrong** [comment]
hi, any updates?

File: example/image-classification/train_model.py (hunk 1 above)
Comment (kernel8liang) [inline example/image-classification/train_model.py]: sorry i just see the message today.

**Comment by kernel8liang** [comment]
@piiswrong sorry, i just read these message today.

File: python/mxnet/optimizer.py (hunk 2 above)
Comment (kernel8liang) [inline python/mxnet/optimizer.py]: done.

File: python/mxnet/optimizer.py [hunk 4]
```diff
@@ -308,6 +308,102 @@ def update(self, index, weight, grad, state):
...
+    rescale_grad : float, optional
+        rescaling factor of gradient.
+
+    clip_gradient : float, optional
+        clip gradient in range [-clip_gradient, clip_gradient]
+
+    param_idx2name : dict of string/int to float, optional
+        special treat weight decay in parameter ends with bias, gamma, and beta
+    """
+    def __init__(self, momentum=0.0, lamda = 0.04, **kwargs):
```
Comment (howard0su) [inline python/mxnet/optimizer.py]: 19:20:27  python/mxnet/optimizer.py:341: convention (C0326, bad-whitespace, ) No space allowed around keyword argument assignment
19:20:27      def __init__(self, momentum=0.0, lamda = 0.04, **kwargs):

Comment (kernel8liang) [inline python/mxnet/optimizer.py]: done.

**Comment by piiswrong** [comment]
please rebase on master to resolve conflicts.
git fetch
git rebase upstream/master

**Comment by kernel8liang** [comment]
done, 
since for a long time, tested again. it works.

**Comment by kernel8liang** [comment]
Hi,

I don't understander this build error,

```
11:09:00 python/mxnet/optimizer.py: 2 Errors of 1 Categories map={'convention': 2}
11:09:00 1 files failed lint
```

is there anything i should do?

**Comment by piiswrong** [comment]
you need to fix code style:
11:08:34  python/mxnet/optimizer.py:356: convention (C0301, line-too-long, ) Line too long (120/100)
11:08:34  python/mxnet/optimizer.py:365: convention (C0301, line-too-long, ) Line too long (123/100)

**Comment by kernel8liang** [comment]
done, should i rebase again?

**Comment by kernel8liang** [comment]
hi,

there is a error in travis-ci, which i don't understander, 

```
Done. Your build exited with 1.
/Users/travis/build.sh: line 148: shell_session_update: command not found
```

is my change case this error?

**Pull request merged**